File: app/src/main/java/com/omarea/common/ui/FastBlurUtility.java
package com.omarea.common.ui;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

public class FastBlurUtility {

    // Tỉ lệ thu nhỏ ảnh để xử lý nhanh (1/10 giúp giảm 100 lần số pixel cần tính toán)
    private static final float SCALE_FACTOR = 0.10f;
    private static final int BLUR_RADIUS = 8;

    /**
     * Chụp màn hình và làm mờ (Dùng làm phương án dự phòng khi không lấy được Wallpaper)
     */
    public static Bitmap getBlurBackgroundDrawer(Activity activity) {
        Bitmap bmp = takeScreenShot(activity);
        if (bmp == null) return null;

        Bitmap blurredBmp = startBlurBackground(bmp);

        // Giải phóng bitmap gốc sau khi đã blur xong
        if (bmp != blurredBmp && !bmp.isRecycled()) {
            bmp.recycle();
        }
        return blurredBmp;
    }

    /**
     * Quy trình xử lý: Thu nhỏ -> Làm mờ -> Phóng to & Nhuộm tối (Dim)
     * Đảm bảo mượt mà và an toàn bộ nhớ từ SDK 23 trở lên.
     */
    public static Bitmap startBlurBackground(Bitmap bkg) {
        if (bkg == null || bkg.isRecycled()) return null;

        int width = Math.round(bkg.getWidth() * SCALE_FACTOR);
        int height = Math.round(bkg.getHeight() * SCALE_FACTOR);

        // Đảm bảo kích thước tối thiểu để không làm hỏng thuật toán
        if (width < 2 || height < 2) return null;

        // 1. Thu nhỏ ảnh (Sử dụng bộ lọc Bilinear để ảnh mượt hơn)
        Bitmap smallBitmap = Bitmap.createScaledBitmap(bkg, width, height, true);

        // 2. Làm mờ bằng thuật toán StackBlur
        Bitmap blurred = fastBlur(smallBitmap, BLUR_RADIUS);

        // Giải phóng smallBitmap tạm thời tránh Memory Leak
        if (smallBitmap != null && smallBitmap != blurred && !smallBitmap.isRecycled()) {
            smallBitmap.recycle();
        }

        if (blurred == null) return null;

        // 3. Phóng to về kích thước gốc và áp dụng bộ lọc màu tối (scaleAndDim sẽ tự recycle 'blurred')
        return scaleAndDim(blurred, bkg.getWidth(), bkg.getHeight());
    }

    /**
     * Chụp ảnh màn hình an toàn trên SDK 23+
     */
    private static Bitmap takeScreenShot(Activity activity) {
        if (activity == null || activity.isFinishing()) return null;
        try {
            View view = activity.getWindow().getDecorView();
            if (view.getWidth() <= 0 || view.getHeight() <= 0) return null;

            Bitmap bitmap = Bitmap.createBitmap(view.getWidth(), view.getHeight(), Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(bitmap);
            view.draw(canvas);
            return bitmap;
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Phóng to ảnh và áp dụng ColorMatrix để làm tối nền (Dim)
     */
    private static Bitmap scaleAndDim(Bitmap bitmap, int targetW, int targetH) {
        if (bitmap == null || bitmap.isRecycled()) return null;

        Bitmap output = Bitmap.createBitmap(targetW, targetH, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(output);

        // Paint với bộ lọc chống răng cưa và lọc bitmap khi scale
        Paint paint = new Paint(Paint.FILTER_BITMAP_FLAG | Paint.ANTI_ALIAS_FLAG);

        // Tạo bộ lọc màu để giảm độ sáng (contrast 0.80f ~ giảm 20% độ sáng)
        ColorMatrix cm = new ColorMatrix();
        float contrast = 0.80f;
        cm.set(new float[]{
                contrast, 0, 0, 0, 0,
                0, contrast, 0, 0, 0,
                0, 0, contrast, 0, 0,
                0, 0, 0, 1, 0});
        paint.setColorFilter(new ColorMatrixColorFilter(cm));

        // Vẽ ảnh từ vùng nguồn (nhỏ) ra vùng đích (toàn màn hình)
        Rect src = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
        Rect dst = new Rect(0, 0, targetW, targetH);
        canvas.drawBitmap(bitmap, src, dst, paint);

        // Giải phóng bitmap tạm sau khi đã vẽ xong
        if (!bitmap.isRecycled()) {
            bitmap.recycle();
        }

        return output;
    }

    /**
     * Thuật toán StackBlur (Multi-pass box blur) - Đã sửa toàn bộ lỗi OutOfBoundsException & Leak
     */
    private static Bitmap fastBlur(Bitmap sentBitmap, int radius) {
        if (sentBitmap == null || sentBitmap.isRecycled()) return null;

        Bitmap.Config config = sentBitmap.getConfig();
        if (config == null) {
            config = Bitmap.Config.ARGB_8888;
        }

        Bitmap bitmap = sentBitmap.copy(config, true);
        if (bitmap == null) return null;

        int w = bitmap.getWidth();
        int h = bitmap.getHeight();
        if (w < 1 || h < 1) return null;

        // Giới hạn bán kính làm mờ không vượt quá kích thước ảnh
        int safeRadius = Math.min(radius, Math.min(w, h) - 1);
        if (safeRadius < 1) safeRadius = 1;

        int[] pix = new int[w * h];
        bitmap.getPixels(pix, 0, w, 0, 0, w, h);

        int wm = w - 1;
        int hm = h - 1;
        int wh = w * h;
        int div = safeRadius + safeRadius + 1;

        int[] r = new int[wh];
        int[] g = new int[wh];
        int[] b = new int[wh];
        int rsum, gsum, bsum, x, y, i, p, yp, yi, yw;
        int[] vmin = new int[Math.max(w, h)];

        int divsum = (div + 1) >> 1;
        divsum *= divsum;
        int dvLimit = 256 * divsum;
        int[] dv = new int[dvLimit];
        for (i = 0; i < dvLimit; i++) {
            dv[i] = (i / divsum);
        }

        yw = yi = 0;
        int[][] stack = new int[div][3];
        int stackpointer;
        int stackstart;
        int[] sir;
        int rbs;
        int r1 = safeRadius + 1;
        int routsum, goutsum, boutsum;
        int rinsum, ginsum, binsum;

        for (y = 0; y < h; y++) {
            rinsum = ginsum = binsum = routsum = goutsum = boutsum = rsum = gsum = bsum = 0;
            for (i = -safeRadius; i <= safeRadius; i++) {
                p = pix[yi + Math.min(wm, Math.max(i, 0))];
                sir = stack[i + safeRadius];
                sir[0] = (p & 0xff0000) >> 16;
                sir[1] = (p & 0x00ff00) >> 8;
                sir[2] = (p & 0x0000ff);
                rbs = r1 - Math.abs(i);
                rsum += sir[0] * rbs;
                gsum += sir[1] * rbs;
                bsum += sir[2] * rbs;
                if (i > 0) {
                    rinsum += sir[0];
                    ginsum += sir[1];
                    binsum += sir[2];
                } else {
                    routsum += sir[0];
                    goutsum += sir[1];
                    boutsum += sir[2];
                }
            }
            stackpointer = safeRadius;

            for (x = 0; x < w; x++) {
                r[yi] = dv[clamp(rsum, 0, dvLimit - 1)];
                g[yi] = dv[clamp(gsum, 0, dvLimit - 1)];
                b[yi] = dv[clamp(bsum, 0, dvLimit - 1)];

                rsum -= routsum;
                gsum -= goutsum;
                bsum -= boutsum;

                stackstart = stackpointer - safeRadius + div;
                sir = stack[stackstart % div];

                routsum -= sir[0];
                goutsum -= sir[1];
                boutsum -= sir[2];

                if (y == 0) vmin[x] = Math.min(x + safeRadius + 1, wm);
                p = pix[yw + vmin[x]];

                sir[0] = (p & 0xff0000) >> 16;
                sir[1] = (p & 0x00ff00) >> 8;
                sir[2] = (p & 0x0000ff);

                rinsum += sir[0];
                ginsum += sir[1];
                binsum += sir[2];

                rsum += rinsum;
                gsum += ginsum;
                bsum += binsum;

                stackpointer = (stackpointer + 1) % div;
                sir = stack[(stackpointer) % div];

                routsum += sir[0];
                goutsum += sir[1];
                boutsum += sir[2];

                rinsum -= sir[0];
                ginsum -= sir[1];
                binsum -= sir[2];

                yi++;
            }
            yw += w;
        }

        for (x = 0; x < w; x++) {
            rinsum = ginsum = binsum = routsum = goutsum = boutsum = rsum = gsum = bsum = 0;
            yp = -safeRadius * w;
            for (i = -safeRadius; i <= safeRadius; i++) {
                yi = Math.min(wh - 1, Math.max(0, yp) + x);
                sir = stack[i + safeRadius];
                sir[0] = r[yi];
                sir[1] = g[yi];
                sir[2] = b[yi];

                rbs = r1 - Math.abs(i);
                rsum += r[yi] * rbs;
                gsum += g[yi] * rbs;
                bsum += b[yi] * rbs;

                if (i > 0) {
                    rinsum += sir[0];
                    ginsum += sir[1];
                    binsum += sir[2];
                } else {
                    routsum += sir[0];
                    goutsum += sir[1];
                    boutsum += sir[2];
                }
                if (i < hm) yp += w;
            }
            yi = x;
            stackpointer = safeRadius;
            for (y = 0; y < h; y++) {
                int rVal = dv[clamp(rsum, 0, dvLimit - 1)];
                int gVal = dv[clamp(gsum, 0, dvLimit - 1)];
                int bVal = dv[clamp(bsum, 0, dvLimit - 1)];

                pix[yi] = (0xff000000 & pix[yi]) | (rVal << 16) | (gVal << 8) | bVal;

                rsum -= routsum;
                gsum -= goutsum;
                bsum -= boutsum;

                stackstart = stackpointer - safeRadius + div;
                sir = stack[stackstart % div];

                routsum -= sir[0];
                goutsum -= sir[1];
                boutsum -= sir[2];

                if (x == 0) vmin[y] = Math.min(y + r1, hm) * w;
                p = Math.min(wh - 1, x + vmin[y]);

                sir[0] = r[p];
                sir[1] = g[p];
                sir[2] = b[p];

                rinsum += sir[0];
                ginsum += sir[1];
                binsum += sir[2];

                rsum += rinsum;
                gsum += ginsum;
                bsum += binsum;

                stackpointer = (stackpointer + 1) % div;
                sir = stack[stackpointer];

                routsum -= sir[0];
                goutsum -= sir[1];
                boutsum -= sir[2];

                rinsum -= sir[0];
                ginsum -= sir[1];
                binsum -= sir[2];

                yi += w;
            }
        }
        bitmap.setPixels(pix, 0, w, 0, 0, w, h);
        return bitmap;
    }

    private static int clamp(int val, int min, int max) {
        return Math.max(min, Math.min(val, max));
    }
}

File: app/src/main/kotlin/com/omarea/common/ui/FastBlurUtility.kt
package com.omarea.common.ui

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.graphics.RectF
import android.view.View
import kotlin.math.max
import kotlin.math.round

/**
 * Tiện ích tạo ảnh nền mờ, dùng cho 2 tình huống RIÊNG BIỆT - KHÔNG dùng chung 1 pipeline
 * như bản cũ, vì bản chất nguồn ảnh khác nhau:
 *
 *   1. VÀO TRANG MỚI (getPageBlurBackground): trang mới chưa có nội dung gì để chụp, nền
 *      mờ ở đây thực chất là ẢNH WALLPAPER đã được BlurController chụp/blur SẴN 1 lần khi
 *      app khởi động / đổi theme (xem BlurController.captureAndBlur, cache trong
 *      BlurEngine.blurBitmap). Chỉ cần scale bitmap cache đó lên full-screen + tint - CHỈ
 *      dùng cache, KHÔNG tự chụp/blur lại (chụp lại mỗi lần mở trang sẽ rất chậm + không
 *      đúng ý nghĩa "ảnh nền" tĩnh theo wallpaper).
 *
 *   2. MỞ DIALOG (getDialogBlurBackground): dialog che lên NỘI DUNG THẬT đang hiển thị của
 *      trang hiện tại (danh sách, text đang gõ dở, v.v...) - ảnh wallpaper cache ở trên
 *      không phản ánh đúng những gì đang thấy trên màn hình. Vì vậy dialog LUÔN chụp
 *      screenshot màn hình activity tại đúng thời điểm mở dialog rồi blur bằng RenderScript
 *      (BlurController) - KHÔNG dùng lại cache wallpaper, đảm bảo nền mờ phía sau dialog
 *      luôn khớp với nội dung thật đang hiển thị ngay trước đó.
 *
 * Cả 2 đều dùng chung RenderScript pipeline của BlurController (GPU-accelerated, ổn định)
 * thay vì StackBlur CPU cũ, và cùng áp dụng contrast thích ứng dark/light mode qua tint.
 *
 * clearCache() phải được gọi khi đổi theme (dark/light hoặc đổi kiểu nền) - ảnh cache cũ đã
 * bake sẵn contrast/tint theo theme CŨ, dùng tiếp sẽ bị sai màu/độ tương phản cho tới khi
 * capture mới xong; xoá cache buộc trang mới phải chờ/for hiện bitmap mới thay vì thấy nhầm
 * ảnh cũ.
 */
object FastBlurUtility {

    /**
     * Ảnh nền mờ cho TRANG MỚI (activity/fragment mới mở) - CHỈ lấy từ cache wallpaper đã
     * blur sẵn (BlurEngine.blurBitmap), KHÔNG tự chụp/blur lại.
     *
     * @return bitmap full-screen đã blur + tint, hoặc null nếu cache chưa sẵn sàng (ví dụ
     *         BlurController vẫn đang capture nền ở background thread - trang nên tạm hiện
     *         nền trống rồi tự cập nhật khi BlurEngine.blurBitmap có giá trị, xem
     *         BlurPreDrawListener). Caller phải recycle bitmap khi không còn dùng.
     */
    @JvmStatic
    fun getPageBlurBackground(activity: Activity): Bitmap? {
        val screenWidth = activity.resources.displayMetrics.widthPixels
        val screenHeight = activity.resources.displayMetrics.heightPixels
        if (screenWidth <= 0 || screenHeight <= 0) return null

        val cachedBlur = BlurEngine.blurBitmap
        if (cachedBlur == null || cachedBlur.isRecycled) return null

        return scaleWithTint(cachedBlur, screenWidth, screenHeight)
    }

    /**
     * Ảnh nền mờ cho DIALOG - LUÔN chụp screenshot màn hình activity hiện tại rồi blur bằng
     * RenderScript, KHÔNG dùng cache wallpaper (BlurEngine.blurBitmap), để nền mờ phía sau
     * dialog khớp đúng nội dung thật đang hiển thị ngay trước khi dialog mở.
     *
     * @return bitmap full-screen đã blur + tint, hoặc null nếu chụp/blur thất bại.
     *         Caller phải recycle bitmap khi không còn dùng.
     */
    @JvmStatic
    fun getDialogBlurBackground(activity: Activity): Bitmap? {
        val screenWidth = activity.resources.displayMetrics.widthPixels
        val screenHeight = activity.resources.displayMetrics.heightPixels
        if (screenWidth <= 0 || screenHeight <= 0) return null

        val screenshot = takeScreenShot(activity)
        if (screenshot == null || screenshot.isRecycled) return null

        val result = blurViaController(activity, screenshot, screenWidth, screenHeight)
        // Screenshot chỉ là nguồn trung gian, recycle sau khi đã dùng xong
        if (!screenshot.isRecycled) {
            screenshot.recycle()
        }
        return result
    }

    /**
     * Xoá cache blur wallpaper (BlurEngine.blurBitmap) - gọi khi đổi theme (dark/light,
     * hoặc bật/tắt directbg) để tránh getPageBlurBackground() trả về ảnh đã bake contrast/
     * tint theo theme CŨ trong lúc chờ BlurController capture xong bản mới.
     *
     * An toàn gọi nhiều lần / gọi khi cache đang null.
     */
    @JvmStatic
    fun clearCache() {
        val cached = BlurEngine.blurBitmap
        if (cached != null && !cached.isRecycled) {
            cached.recycle()
        }
        BlurEngine.blurBitmap = null
    }

    /**
     * Blur trực tiếp 1 bitmap đã chụp sẵn từ nơi khác (không tự chụp screenshot của activity).
     *
     * Dùng cho SwipeBackPreviewCache: bitmap "sharp" đã được chụp qua PixelCopy (giữ đúng bo
     * góc/clip) - chỉ cần blur nó bằng pipeline RenderScript sẵn có, giữ nguyên kích thước gốc
     * của bitmap đầu vào.
     *
     * @return bitmap đã blur + tint (cùng kích thước với sourceBitmap), hoặc null nếu lỗi.
     *         Caller phải recycle bitmap khi không còn dùng.
     */
    @JvmStatic
    fun blurBitmap(activity: Activity, sourceBitmap: Bitmap): Bitmap? {
        if (sourceBitmap.isRecycled) return null
        return blurViaController(activity, sourceBitmap, sourceBitmap.width, sourceBitmap.height)
    }

    /**
     * Chụp ảnh màn hình an toàn.
     */
    private fun takeScreenShot(activity: Activity): Bitmap? {
        return try {
            val view: View = activity.window.decorView
            if (view.width <= 0 || view.height <= 0) return null

            val bitmap = Bitmap.createBitmap(view.width, view.height, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            view.draw(canvas)
            bitmap
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Scale bitmap blur (nhỏ, ~20% screen) lên full-screen và phủ tint.
     *
     * Dùng cho cả 2 trường hợp:
     *   - Bitmap cache từ BlurEngine.blurBitmap
     *   - Bitmap blur vừa tạo mới
     *
     * @return bitmap mới full-screen, hoặc null nếu lỗi. Caller phải recycle.
     */
    private fun scaleWithTint(blurBitmap: Bitmap, targetW: Int, targetH: Int): Bitmap? {
        return try {
            val output = Bitmap.createBitmap(targetW, targetH, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(output)

            val paint = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)

            // Áp dụng contrast + tint giống BlurController
            val contrastValue = if (com.tool.tree.ThemeModeState.isDarkMode()) 0.9f else 1.2f
            val offset = (1f - contrastValue) * 128f
            val cm = ColorMatrix(
                floatArrayOf(
                    contrastValue, 0f, 0f, 0f, offset,
                    0f, contrastValue, 0f, 0f, offset,
                    0f, 0f, contrastValue, 0f, offset,
                    0f, 0f, 0f, 1f, 0f
                )
            )
            paint.colorFilter = ColorMatrixColorFilter(cm)

            canvas.drawBitmap(blurBitmap, null, RectF(0f, 0f, targetW.toFloat(), targetH.toFloat()), paint)
            output
        } catch (e: OutOfMemoryError) {
            null
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Blur screenshot bằng RenderScript thông qua BlurController.
     *
     * Thay vì dùng StackBlur CPU (bản cũ, hay fail), gọi trực tiếp vào
     * BlurController để tận dụng:
     *   - RenderScript GPU-accelerated (nhanh + ổn định hơn)
     *   - RS context/script đã cache (không tốn thời gian tạo lại)
     *   - Scale 20% + radius 16f (chất lượng cao hơn bản cũ)
     *
     * @return bitmap full-screen đã blur + tint, hoặc null. Caller phải recycle.
     */
    private fun blurViaController(activity: Activity, screenshot: Bitmap, screenWidth: Int, screenHeight: Int): Bitmap? {
        return try {
            // Scale xuống 20% giống BlurController (nhanh + đủ chất lượng)
            val scale = 0.20f
            val width = max(round(screenshot.width * scale).toInt(), 1)
            val height = max(round(screenshot.height * scale).toInt(), 1)
            val scaled = Bitmap.createScaledBitmap(screenshot, width, height, true)

            // Blur bằng RenderScript (BlurController.cacheBlurBitmap)
            val blurred = BlurEngine.controller.cacheBlurBitmap(activity.applicationContext, scaled, 16f)

            // scaled đã dùng xong, recycle
            if (!scaled.isRecycled) {
                scaled.recycle()
            }

            if (blurred == null || blurred.isRecycled) return null

            // Scale lên full-screen + tint
            val result = scaleWithTint(blurred, screenWidth, screenHeight)

            // blurred là bitmap trung gian, recycle (trừ khi nó được gán vào BlurEngine.blurBitmap bên trong controller)
            // An toàn: chỉ recycle nếu không phải là bitmap đang được cache
            if (blurred !== BlurEngine.blurBitmap && !blurred.isRecycled) {
                blurred.recycle()
            }

            result
        } catch (e: OutOfMemoryError) {
            null
        } catch (e: Exception) {
            null
        }
    }
}
